### ISSUE TITLE
Multilingual-MiniLM-L12-H384 ONNX inference in NodeJS

### ISSUE THREAD

**Issue by loretoparisi** [microsoft/onnxruntime#13171]
### Describe the issue

I have converted a fine-tuned [Multilingual-MiniLM-L12-H384](https://huggingface.co/microsoft/Multilingual-MiniLM-L12-H384) to ONNX. I have also successfully converted the `sentencepiece.bpe.model` tokenizer using HuggingFace tokenizers library. See [here](https://github.com/huggingface/tokenizers/issues/1076) for more details about this, generating the JSON vocab file and merge file:

```javascript
const sentencePieceBertTokenizer = await SentencePieceBPETokenizer.fromOptions({
        vocabFile: "sentencepiece-vocab.json",
        mergesFile: "sentencepiece-merges.txt",
    });
```

The tokenizer properly works  and generated tokens and encoded ids:

```
    let encoder = (tokenizer) => promisify(tokenizer.encode.bind(tokenizer))
    encode = encoder(tokenizer);
    
    encoded = await encode("Hello how are you?");
    console.log("ids ", encoded.getIds());
    console.log("tokens ", encoded.getTokens());
    decoded = await tokenizer.decode(encoded.getIds(), skipSpecialTokens);
    console.log(decoded); // hello how are you?
getting the expected result

ids  [ 35377, 3641, 620, 397, 31 ]
tokens  [ '▁Hello', '▁how', '▁are', '▁you', '?' ]
Hello how are you?
```

### To reproduce

When passing the generated ids to the model inference in ONNX:

```javascript
const ort = require('onnxruntime-node');

// onnx sesion
session = await ort.InferenceSession.create(self._options.model.path, options);

const encoded_ids = (await tokenizer.tokenize(text)).getIds();
const model_input = ONNX.create_model_input(encoded_ids);
const output = await session.run(model_input, ['output_0']);
```

I get an invalid expand shape error

```
2022-09-29 18:59:08.039 node[63957:683481] 2022-09-29 17:59:08.039329 [E:onnxruntime:, parallel_executor.cc:210 RunNodeAsync] Non-zero status code returned while running Expand node. Name:'Expand_21' Status Message: invalid expand shape
2022-09-29 18:59:08.040 node[63957:683429] 2022-09-29 17:59:08.039978 [E:onnxruntime:, parallel_executor.cc:75 Execute] [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running Expand node. Name:'Expand_21' Status Message: invalid expand shape
```

Model input was

```
model_input  {
  input_ids: l {
    dims: [ 1, 604 ],
    type: 'int64',
    data: BigInt64Array(604) [
        101n,     5n,     0n,   567n,   807n,  4504n,   91n,   397n,
       6164n,    75n,  1379n,   644n,     0n, 69084n, 1365n,    86n,
      49781n,   441n,  5674n,   515n, 58967n,    69n, 4223n,  1080n,
      34270n,     0n,   283n,   106n,   397n,  3567n, 3677n,    31n,
      15900n,   397n,   220n,  1296n,    31n,     0n,  131n,   617n,
       4504n,    91n,  1671n,   441n,    15n,     0n, 8330n,    89n,
         18n,    24n,    17n, 26865n,     0n, 37838n, 2173n,   441n,
      30481n,   397n, 12318n,  4126n,  7067n,    86n, 5153n,    53n,
        441n,     0n,   567n,  2300n,    24n,    17n, 3713n,  2366n,
        397n,    24n,   106n, 89288n,    99n,     0n,  918n, 11632n,
        758n, 79196n,   256n,  3407n,  9587n,   213n,    0n, 23388n,
         24n,     6n,  5791n,  5044n, 80096n,     0n,  567n,    24n,
         38n,   705n,   190n,    75n,
      ... 504 more items
    ],
    size: 604
  },
  attention_mask: l {
    dims: [ 1, 604 ],
    type: 'int64',
    data: BigInt64Array(604) [
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n, 1n,
      1n, 1n, 1n, 1n,
      ... 504 more items
    ],
    size: 604
  },
  token_type_ids: l {
    dims: [ 1, 604 ],
    type: 'int64',
    data: BigInt64Array(604) [
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n, 0n,
      0n, 0n, 0n, 0n,
      ... 504 more items
    ],
    size: 604
  }
}
```

### Platform

Mac

### OS Version

macOS 10.14.5

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

onnxruntime-node@1.11.0

### ONNX Runtime API

JavaScript

### Architecture

X86

### Execution Provider

Default CPU

### Execution Provider Library Version

-

**Comment by fs-eire** [comment]
I don't see anything wrong with the JavaScript code, and the inputs looks good.

The error message shows that the input shape is invalid for a 'Expand' node in the model. probably a problem with the converter or the ort core.

**Comment by loretoparisi** [comment]
> I don't see anything wrong with the JavaScript code, and the inputs looks good.
> 
> The error message shows that the input shape is invalid for a 'Expand' node in the model. probably a problem with the converter or the ort core.

@fs-eire thank you for the hint, @paulthemagno please let's have a look at the converted model!

**Comment by loretoparisi** [comment]
> I don't see anything wrong with the JavaScript code, and the inputs looks good.
> 
> The error message shows that the input shape is invalid for a 'Expand' node in the model. probably a problem with the converter or the ort core.

@fs-eire we digged into the converted model a bit, and it may happen that  there is a difference in the special tokens of the original model tokenizer and converted one (likely the `<pad>` and the `<mask>` are missing in the converted one). In fact we see a difference in the size of the vocab (25000 against 250002 where 2 is accounted for the missing special tokens). If so, could be the `Expand` error related?

Thank you!

**Comment by fs-eire** [comment]
It's possible.

There is one more thing can do to test: use the onnxruntime python package to perform the model inference using the same input tensor data. If it runs successfully, then there must be something wrong with Node.js binding; otherwise, it's the problem of either ORT core or the model itself. Then, need the model file for further investigation.

**Comment by loretoparisi** [comment]
> It's possible.
> 
> There is one more thing can do to test: use the onnxruntime python package to perform the model inference using the same input tensor data. If it runs successfully, then there must be something wrong with Node.js binding; otherwise, it's the problem of either ORT core or the model itself. Then, need the model file for further investigation.

@fs-eire Thanks a lot @paulthemagno let's try the python counterpart.

**Comment by paulthemagno** [comment]
I tried to convert the pytorch model with this script:

```python
import torch
from transformers import AutoModelForSequenceClassification

def convert_pt_to_onnx(input_pt_path, output_onnx_path):
    model = AutoModelForSequenceClassification.from_pretrained(input_pt_path)
    dummy_input_ids = torch.Tensor([[0, 3293, 83, 142, 98, 19, 425, 3299, 2]]).to(torch.int64)
    torch.onnx.export(model,
                  (dummy_input_ids),
                  output_onnx_path,
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=11,          # the ONNX version to export the model to
                  do_constant_folding=True,  # whether to execute constant folding for optimization
                  input_names=['input'],
                  output_names=['output'],
                  dynamic_axes={'input' :{0 : 'batch_size',
                                          1: 'sentence_length'},
                                'output': {0: 'batch_size'}})

convert_pt_to_onnx(input_pt_path = "MY_PT_MODEL_PATH", output_onnx_path ="ONNX_PATH.onnx")
```

Then I load the session with this function:
```python
import torch
from os import environ
from psutil import cpu_count
from transformers import AutoTokenizer
import onnx
from onnxruntime import GraphOptimizationLevel, InferenceSession, SessionOptions, get_all_providers

def create_inference_session(model_path: str, use_gpu: bool) -> InferenceSession: 
    # Constants from the performance optimization available in onnxruntime
    # It needs to be done before importing onnxruntime
    environ["OMP_NUM_THREADS"] = str(cpu_count(logical=True))
    environ["OMP_WAIT_POLICY"] = 'ACTIVE'
    PROVIDERS = ["CPUExecutionProvider", "CUDAExecutionProvider"]
    
    if torch.cuda.is_available() and use_gpu:
        provider = PROVIDERS[1] #cuda
    else: 
        provider = PROVIDERS[0] #cpu
    print(f"Using provider {provider}")
  
    assert provider in get_all_providers(), f"provider {provider} not found, {get_all_providers()}"

    # Few properties that might have an impact on performances (provided by MS)
    options = SessionOptions()
    options.intra_op_num_threads = 1
    options.graph_optimization_level = GraphOptimizationLevel.ORT_ENABLE_ALL

    # Load the model as a graph and prepare the CPU backend 
    session = InferenceSession(model_path, options, providers=[provider])
    session.disable_fallback()

    return session

onnx_path = "ONNX_PATH.onnx"
tokenizer = AutoTokenizer.from_pretrained('microsoft/Multilingual-MiniLM-L12-H384')
ort_session = create_inference_session(model_path = onnx_path, use_gpu = True)
```

And finally I defined the predict function:

```python
import torch
import numpy as np

def predict_batch(ort_session, tokenizer, texts, labels):
    def to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
    input_ids = tokenizer.batch_encode_plus(texts, add_special_tokens=True, return_token_type_ids=False, return_attention_mask=False)["input_ids"]
    
    # compute ONNX Runtime output prediction
    outputs = []
    logits = []
    for input_id in input_ids:
        input_id = torch.tensor(input_id)
        input_id = torch.unsqueeze(input_id, dim=0)
        if max_length is not None and len(input_id) > max_length:
            input_id = input_id[:max_length]
        ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(input_id)}
        ort_out = ort_session.run(None, ort_inputs)
        sigmoid_tensor = torch.sigmoid(torch.tensor(np.array(ort_out)))
        int_label = torch.argmax(sigmoid_tensor)
        logits.append(np.squeeze(sigmoid_tensor.cpu().detach().numpy()))
        outputs.append(labels[int_label])
    return {"logits": logits, "labels": outputs}
```

It's working even if I find different results between the execution of the ONNX model and the Pytorch one (labels are correct but score is slightly different)

```python
# with ONNX
print(predict_batch(ort_session = ort_session,
                    tokenizer = tokenizer, 
                    texts = [
                        "Example1",
                        "Example2",
                        "Example3",    
                    ], 
                    labels = ["LABEL_0", "LABEL_1"]))

# with Pytorch model
model = AutoModelForSequenceClassification.from_pretrained("MY_PT_MODEL_PATH")
tokenizer = AutoTokenizer.from_pretrained('microsoft/Multilingual-MiniLM-L12-H384')
classifier = pipeline('text-classification', model=model, tokenizer=tokenizer, device = 0)
print(classifier([
    "Example1",
    "Example2",
    "Example3",    
]))
```

and the ouptut is similar, but not the same:

```
//onnx
{'logits': [
array([0.96897584, 0.02641451], dtype=float32),
 array([0.97955716, 0.01729073], dtype=float32), 
array([0.95629483, 0.0375994 ], dtype=float32)], 
'labels': [
'LABEL_0', 
'LABEL_0', 
'LABEL_0']
}

//pytorch
[{'label': 'LABEL_0', 'score': 0.9991320967674255},
 {'label': 'LABEL_0', 'score': 0.9996329545974731},
 {'label': 'LABEL_0', 'score': 0.9982176423072815}]
```
